crates/hir-ty/src/infer/closure.rs

```diff
@@ -424,6 +427,39 @@ impl InferenceContext<'_> {
         )))
     }
 
+    fn extract_async_fn_sig_from_projection(
+        &mut self,
+        projection_ty: &ProjectionTy,
+        projected_ty: &Ty,
+    ) -> Option<FnSubst<Interner>> {
+        let arg_param_ty = projection_ty.substitution.as_slice(Interner)[1].assert_ty_ref(Interner);
+
+        let TyKind::Tuple(_, input_tys) = arg_param_ty.kind(Interner) else {
+            return None;
+        };
```

Comment:
dumb question: under what circumstances would we not get a `TyKind::Tuple` here?

Reply:
I don't know but I can guess it can happen if users are messing with the raw `fn_traits` unstable feature. This was in `extract_sig_from_projection()` so I duplicated it.